In [1]:
#import libraries
import pandas as pd
import numpy as np
import os


In [2]:
path = r'/Users/ethanahennes/Desktop/instacart basket analysis'

In [9]:
df_ords = pd.read_csv(os.path.join(path,'02 data','Original Data','4.3_orders_products','orders.csv'), index_col = False)

In [10]:
df_prods = pd.read_csv(os.path.join(path,'02 data','Original Data','4.3_orders_products','products.csv'), index_col = False)

In [12]:
#dropping eval_set column
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [13]:
# dropping eval_set from df_ords
df_ords = df_ords.drop(columns = ['eval_set'])

In [18]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [21]:
#changing the name of a column
df_ords.rename(columns = {'order_dow' :'orders_day_of_week'}, inplace = True)

In [22]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [24]:
# changing a variables data type so it is ignornd when gathering descriptive stats
df_ords['order_id']=df_ords['order_id'].astype('str')

In [25]:
df_ords['order_id'].dtype

dtype('O')

In [33]:
df_dep = pd.read_csv(os.path.join(path,'02 data','Original Data','4.3_orders_products','departments.csv'), index_col = False)

In [34]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [35]:
# transposong columns to rows
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [36]:
# creating a new df_dep to make perminatnt
df_dep_t = df_dep.T

In [38]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [39]:
#adding and index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

#in order to remove the '0' we have to make a copy of our first row

In [41]:
new_header = df_dep_t.iloc[0]

In [42]:
new_header

0    department
Name: department_id, dtype: object

#then we detle the row 

In [45]:
df_dep_t_new=df_dep_t[1:]

In [46]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

#then we re fuse it again?

In [47]:
df_dep_t_new.columns = new_header

In [48]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [49]:
data_dict = df_dep_t_new.to_dict('index')

In [50]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [51]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [52]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [53]:
df_snacks = df_prods[df_prods['department_id']==19]

In [54]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [55]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [58]:
df_snacks = df_prods[df_prods['department_id']==19]

In [59]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [60]:
df_snacks_2 = df_prods.loc[df_prods['department_id'].isin([19])]

# TASK

In [62]:
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [63]:
# 2)changing user_id to string
df_ords['user_id'] = df_ords['user_id']. astype('str')

In [65]:
#checking to see if type has changed
df_ords['user_id'].dtype

dtype('O')

In [66]:
# 3) looking for an unituitive name in the df_ords
df_ords.rename(columns = {'days_since_prior_order' :'days_from_last_order'}, inplace = True)

In [67]:
df_ords

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_from_last_order  
0                         NaN  
1                        15.0  
2                        21.0  
3                        29.0  
4                        28.0  
...                       ...  
3421078                  29.0  
3421079                  30.0  
3421080                  18.0  
3421081                   7.0  
3421082                  30.0  

[3421083 rows x 6 columns]

In [68]:
#making a new table
df_ords_II = df_ords.rename(columns = {'days_since_prior_order' :'days_from_last_order'}, inplace = True)

In [70]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_from_last_order  
count          3.214874e+06  
mean           1.111484e+01  
std            9.206737e+00  
min            0.000000e+00  
25%            4.000000e+00  
50%            7.000000e+00  
75%            1.500000e+01  
max            3.000000e+01

In [72]:
# 4) finding the busiest hour of the day
df_ords.groupby('order_hour_of_day').size()

order_hour_of_day
0      22758
1      12398
2       7539
3       5474
4       5527
5       9569
6      30529
7      91868
8     178201
9     257812
10    288418
11    284728
12    272841
13    277999
14    283042
15    283639
16    272553
17    228795
18    182912
19    140569
20    104292
21     78109
22     61468
23     40043
dtype: int64

#the busiest hour of the day is 10am 

In [73]:
# 5) fidning the value behind 4 in the "department_id"
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

#the value behind 4 is 'produce'

In [74]:
# 6) more information breakfat items
print(data_dict.get('14'))

{'department': 'breakfast'}


In [75]:
df_breakfast = df_prods[df_prods['department_id']==14]

In [76]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [97]:
#7 find information about alchohol,deli,beverages, and meat/seafood (5,20,7,12)
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [98]:
#checking the success of subset
df_dinner_parties.tail()

product_id           product_name  aisle_id  department_id  prices
49425       49421       Salame, Peppered       106             12    12.2
49440       49436  Imitation Crab Flakes        15             12    23.5
49509       49505    Hot Italian Sausage       106             12    15.8
49655       49651           Beef Brisket       122             12    20.7
49674       49670    Beef Summer Sausage       106             12    19.2

In [99]:
#looking for the most ordered within the subset
df_dinner_parties.groupby('department_id').size()

department_id
5     1056
7     4365
12     907
20    1322
dtype: int64

In [100]:
Df_dinner_parties

product_id                                       product_name  \
16             17                                  Rendered Duck Fat   
22             23                             Organic Turkey Burgers   
34             35     Italian Herb Porcini Mushrooms Chicken Sausage   
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
...           ...                                                ...   
49425       49421                                   Salame, Peppered   
49440       49436                              Imitation Crab Flakes   
49509       49505                                Hot Italian Sausage   
49655       49651                                       Beef Brisket   
49674       49670                                Beef Summer Sausage   

       aisle_id  department_id  prices  
16           35             12    17.1  
22           49             12     8.2  
34          106             12    15.1  
39          106             12    22.5  
83            7             12    24.3  
...         ...            ...     ...  
49425       106             12    12.2  
49440        15             12    23.5  
49509       106             12    15.8  
49655       122             12    20.7  
49674       106             12    19.2  

[907 rows x 5 columns]

#8) the data frame 'df_dinner_parties' has 907 rows and 5 columns

In [105]:
# 9) creating at df for user_id 1
df_user_1=df_ords[df_ords['user_id']=='1']

In [106]:
df_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_from_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0  
5                   19.0  
6                   20.0  
7                   14.0  
8                    0.0  
9                   30.0  
10                  14.0

In [111]:
#grouping user_1 by the day of the week they order and at what time of day they order on that day
df_user_1.groupby(['orders_day_of_week','order_hour_of_day']).size()

orders_day_of_week  order_hour_of_day
1                   9                    1
                    14                   1
                    16                   1
2                   7                    1
                    8                    1
3                   7                    1
                    12                   1
4                   7                    1
                    8                    2
                    15                   1
dtype: int64

In [113]:
df_ords.to_csv(os.path.join(path,'02 data','Pepared Data','orders_wrangled.csv))'))

In [114]:
df_dep_t_new.to_csv(os.path.join(path,'02 data','Pepared Data','new_dep_wrangled.csv))'))